김호집/PROGRAMMERS/java/Week15/두큐합같게만들기.java

```diff
@@ -0,0 +1,128 @@
+/**
+ * Author    : HojipKim
+ * Date      : 2024.09.07(Sat)
+ * Runtime   : 1681 ms
+ * Memory    : 409 MB
+ * Algorithm : Queue
+ */
+
+import java.util.Queue;
+import java.util.ArrayDeque;
+
+class Main{
+    public static void main(String[] args) {
+        Solution solution = new Solution();
+        int ans = solution.solution(new int[]{1, 1}, new int[]{1, 5});
+        System.out.println(ans);
+
+    }
+}
+
+class Solution {
+
+    public int solution(int[] queue1, int[] queue2) {
+        int answer = 0;
+
+        long sum = addAllElementTwoArray(queue1, queue2);
+
+        Queue<Integer> q1 = new ArrayDeque<>();
+        Queue<Integer> q2 = new ArrayDeque<>();
+
+        if(sum == -1 || !isEvenNumber(sum)){
+            return -1;
+        }
+
+        long sum1 = addAllElement(queue1);
+        long sum2 = addAllElement(queue2);
+        if(sum1 + sum2 == 0){
+            return 0;
+        }
+
+        int q2Res = 0;
+        if(!(arrayToQueue(queue1, q1) && arrayToQueue(queue2, q2))){
+            return -1;
+        }
+
+        q2Res = startOrderByQ2(q1, q2, sum1, sum2);
+
+        if(q2Res == -1){
+            return -1;
+        }
+        answer = q2Res;
+
+        return answer;
+    }
+
+    public static int startOrderByQ2(Queue<Integer> q1, Queue<Integer> q2, long sum1, long sum2){
+        int cnt = 0;
+
+        while(sum1 != sum2){
+            if(cnt > 100_000_000){
+                return -1;
+            }
+            if(sum1 > sum2){
+                int tmp = q1.poll();
+                q2.offer(tmp);
+                sum1 -= tmp;
+                sum2 += tmp;
+                cnt++;
+            }else if(sum1 < sum2){
+                int tmp = q2.poll();
+                q1.offer(tmp);
+                sum1 += tmp;
+                sum2 -= tmp;
+                cnt++;
+            }else{
+                break;
+            }
+
+        }
+        return cnt;
+    }
+
+    public static long addAllElement(int[] arr){
+        long sum = 0;
+
+        for(long element : arr){
+            sum += element;
+        }
+
+        return sum;
+    }
+
+    public static boolean arrayToQueue(int[] arr, Queue<Integer> queue){
+        for(int elem : arr){
+            queue.offer(elem);
+        }
+        if(queue.size() == arr.length){
+            return true;
+        }
+
+        return false;
```

Comment:
```suggestion
        return queue.size() == arr.length;
```

이렇게 코드 구성해도 괜찮아 보이네요.

Reply:
코드를 확인해 보니까 해당 메소드의 존재 의미가 잘 이해가 안가네요.

메소드가 호출되는 시점에 Queue 는 초기화 상태로 인스턴스 생성만 된 상태로 존재(= 빈 값)
arr 요소를 Queue 로 넣음(= arr과 Queue 는 같은 요소를 갖음)

```
return arr.length = arr.length;
```

위와 같은 코드가 아닌가 싶어서 혼란 스럽네요. @Hojip-Kim

Reply:
맞습니다.
사실 습관적으로 유효성 체크를해준다는것이 굳이 불필요한곳에까지 체크를하게되었네요.

Reply:
유효성 검사는 언제나 좋으니까 유효한 검증 코드 작성 습관은 좋은 것 같네요.